Spring Web으로 ToDo REST API 만들기

alex-gibeom opened the pull request:
이번 한주도 잘 부탁드립니다 :)
열심히 배우며 따라가겠습니다!

alex-gibeom committed:
feat: Task 목록 조회 및 상세 조회 API

- Service 생성
- Repository 생성
- ReqeustDto 생성

alex-gibeom committed:
merge: [Day1] - Spring으로 ToDo REST API 만들기

alex-gibeom commented:
오늘 예비군 갔다와서 많이 못했네요ㅠ
계속 추가적으로 올리겠습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/repository/TaskRepository.java:
```diff
@@ -0,0 +1,28 @@
+        return tasks;
+    }
+
+    public Task getTaskById(TaskSearchDto taskSearchDto) {
```
`TaskSearchDto` 에 `title` 은 사용하지 않는 것일까요?

alex-gibeom replied:
요구 사항에서는 id를 통해 상세 조회를 하는 것으로 인지하였습니다!
(309ada0) 따라서 해당 `TaskSearchDto` 는 제거하고 실제 PathVariable인 id를 그대로 service로 넘기도록 수정하였습니다!

ssisksl77 replied:
아주 좋아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/controller/TaskController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @GetMapping("/{id}")
+    public Task get(@PathVariable Long id) {
+        TaskSearchDto taskSearchDto = new TaskSearchDto();
+        taskSearchDto.setId(id);
```
초기화는 한번에 수행하는 것이 좋습니다. 이펙티브자바를 참고해주세요! 관련 아이템을 읽고 공유해주세요!

alex-gibeom replied:
어떤 아이템인지 파악이 잘 안돼서, 일단 제 판단으로 아래의 주제들을 공부해보았습니다!

- 아이템 17. 변경 가능성을 최소화 하라
- 아이템 2. 생성자에 매개변수가 많다면 빌더를 고려하라

[변경 가능성을 최소화 하라]
- 생성자는 불변식 설정이 모두 완료된, 초기화가 완벽히 끝난 상태의 객체를 생성해야 한다.
- 다른 합당한 이유가 없다면 모든 필드는 private final이어야 한다.
- 객체의 상태를 변경하는 메서드(변경자)를 제공하지 않는다 -> setter 금지
- 클래스를 확장할 수 없도록 한다 -> 이건 잘 이해가 안됩니다😂

[생성자에 매개변수가 많다면 빌더를 고려하라]
- 객체 생성 시, 매개변수를 직관적으로 세팅할 수 있으며 가독성이 좋다.
- 쓰레드 안전성이 우수하다.


위의 주제를 공부해보는 것을 의도하신걸로 인지하였습니다..!
아니였다면 다시 말씀해주시면 다시 공부해오겠습니다 :)
(여전히 어려운 책이네요..! 😂)

ssisksl77 replied:
너무 좋습니다!

ssisksl77 replied:
특히 지금 사용하신 방식은 이팩티브 자바에서는 자바빈패턴이라고 말하죠! setter를 이용한 초기화는 언제 제대로 된 객체가 되는 것인지 인지를 못하게 될 수 있습니다. 가령
```java
Order order = new Order();
order.setId();
order.setStatus(,,,);

// 뭔가 일을 더함.
order.setPG();
```

이런식의 코드가 있다면 언제부터 제대로된 Order 객체인지 모르게되겠죠!

alex-gibeom replied:
멘토님의 말씀처럼 setter를 이용하게되면 "불변객체"가 성립되지 않아서 객체 데이터에 대한 신뢰가 안 생길 것 같다고 생각되었습니다.
(309ada0) 따라서 모든 DTO와 Entity 객체의 setter 메서드를 삭제하고, Builder를 통해 생성할 수 있도록 수정해보았습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+
+@Service
+public class TaskServiceImpl implements TaskService{
```
```suggestion
public class TaskServiceImpl implements TaskService {
```
nit : 아주 사소하지만 일관성있게 정리된 것이 낫죠!

alex-gibeom replied:
아 맞습니다!
어제 급하게 PR 올리려고 부랴부랴 코딩하느라 놓쳤나보네요..!
감사합니다 :)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+public class TaskServiceImpl implements TaskService{
+    private final TaskRepository taskRepository;
+
+    @Autowired
```
`@Autowired` 가 없이도 동작하는지 확인해보아요!

alex-gibeom replied:
`@Autowired`가 없어도 잘 동작 되네요 🙂

찾아보니 `@Autowired`를 생략하여도 **생성자가 단 하나**라면, 암시적으로 주입을 시켜주는 조건이 있는 것을 알게 되었습니다!
TaskServiceImpl 생성자가 하나여서 생략되어도 되는 것 같습니다 :)

참고자료: [baeldung.com](https://www.baeldung.com/constructor-injection-in-spring#implicit-constructor-injection)
> 3. Implicit Constructor Injection
As of Spring 4.3, classes with a single constructor can omit the @Autowired annotation. This is a nice little bit of convenience and boilerplate removal.
On top of that, also starting with 4.3, we can leverage the constructor-based injection in @Configuration annotated classes. In addition, if such a class has only one constructor, we can omit the @Autowired annotation as well.

ssisksl77 replied:
아주 좋습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/repository/TaskRepository.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public Task getTaskById(TaskSearchDto taskSearchDto) {
+        if (taskSearchDto.getId() == null) {
```
더 생각해보면 taskSerachDto 도 null일 수 있지 않을까요?

alex-gibeom replied:
해당 코드에서는 taskSearchDto가 null일 확률도 있다고 생각합니다!
(309ada0) 변경한 코드에서는 @PathVariable을 통해 받아오는 id를 그대로 service로 전달해주기 때문에 id가 null로 들어올 경우 일반 목록 조회 컨트롤러로 매핑됩니다!

alex-gibeom committed:
feat: Spring devtool 추가

alex-gibeom committed:
feat: Lombok 라이브러리 추가

alex-gibeom committed:
feat: RequestDto, ResponseDto, ResponseEntity 분리 작업

- 상세 조회는 PathVariable로 진행하기 때문에 매개변수 값 그대로 Service로 전달
   - SearchDto 삭제

alex-gibeom committed:
feat: ToDo 생성하기 API

alex-gibeom committed:
feat: ToDo 수정하기 API

alex-gibeom committed:
merge: [Day2] - Spring으로 ToDo REST API 만들기

alex-gibeom commented:
추가로 오늘 작업건에 대해 아래와 같이 설계를 해보았습니다!
피드백 해주시면 감사하겠습니다 😄

![image](https://user-images.githubusercontent.com/59248326/196424732-ffdc0e17-5434-495f-b316-b1c9bd22b4ca.png)

alex-gibeom committed:
feat: ToDo 부분 수정하기 API

alex-gibeom committed:
fix: Todo 생성하기 API 성공 응답코드 수정 (200 -> 201)

alex-gibeom committed:
feat: ToDo 삭제하기 API

alex-gibeom committed:
feat: 공통 Exception 추가 및 각 API Error 상황 적용

alex-gibeom committed:
feat: Spring Boot Validation 의존성 추가

alex-gibeom committed:
feat: requestBody 데이터(title) validation 추가 - 빈 문자 허용 안함

alex-gibeom committed:
feat: Validation Exception 핸들링 추가

alex-gibeom committed:
feat: PathVariable id 제약 조건 추가 - 0 이상

alex-gibeom committed:
refactor: ToDo 상세 조회 Repository 반환 타입 Optional 처리

alex-gibeom committed:
refactor: 기존 Repository 내부 반환 값 검증 로직(Exception) 모두 Service로 응집

- Repository 역할 : 순수 데이터 처리
- Service 역할
   - RequestDTO -> Entity 변환
   - 반환 데이터에 필요한 데이터 로직 수행
   - 데이터 처리 로직 중 반환 값 검증 수행
- Controller 역할
   - Request 데이터 검증
   - Request 데이터 모두 Service로 전달
   - 응답 Entity -> ResponseDTO 변환

alex-gibeom committed:
merge:  [Day3] - 백엔드 구현 완료

alex-gibeom commented:
각 Controller, Service, Repository의 역할을 정리해 보았습니다!
피드백 해주시면 감사하겠습니다 😄

#### [Controller 역할]
- PathVariable 데이터 검증
- RequestBody 데이터 검증
- 요청 데이터 Service로 전달
- Service로부터 반환된 Entity -> ResponseDto 변환 및 응답 반환

#### [Service 역할]
- RequestDto -> Entity 변환
- 응답 데이터에 필요한 데이터 로직 수행
- 데이터 로직 도중 반환 값 검증 수행
- Controller로 Entity 반환

#### [Repository 역할]
- 넘어온 Entity를 통해 단순 데이터 CRUD 작업 후 데이터 반환

alex-gibeom commented:
Exception 핸들링 Flow를 정리해보았습니다!
피드백 해주시면 감사하겠습니다 😄

#### [Exception 핸들링 Flow]

<img src="https://user-images.githubusercontent.com/59248326/196682448-cbb50e21-9619-4cc8-be10-c3389607587c.png" width="100%">

#### [각 핸들러 역할]
**MethodArgumentNotValidException Handler**
- @RequestBody로 받는 Dto의 Valid 검증 실패 핸들러 (ArgumentResolver 구현체)

**ConstraintViolationException Handler**
- @PathVariable의 Valid 검증 실패 핸들러
- 클래스 단위에서 @Validated 사용 시 모든 메서드에서 사용 가능 (인터셉터)

**RestApiException Handler**
- ErrorCode enum에서 직접 선언한 커스텀 에러 핸들러 (404, 409 에러)

hannut91 commented on app/build.gradle:
```diff
@@ -36,6 +43,24 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
+
+/*  Spring Boot Validation
+    - Spring 2.3 부터는 spring-boot-starter-web 의존성 내부에 있던 validation이 사라짐
```
```suggestion
    - Spring Boot 2.3 부터는 spring-boot-starter-web 의존성 내부에 있던 validation이 사라짐
```

Spring Boot를 말씀하시고 싶으셨던 것 같습니다

alex-gibeom replied:
아 맞습니다!
Spring이랑 Spring Boot의 버전 차이가 많이 다른데, 혼동되지 않게 주의해야겠습니다.
감사합니다 :)

hannut91 commented on app/src/main/java/com/codesoom/assignment/IdGenerator.java:
```diff
@@ -0,0 +1,12 @@
+    }
+
+    public static Long createId() {
+        return ++id;
```
전위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 ++value 와 value++ 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

alex-gibeom replied:
타인이 봤을 때 조금이라도 생각이 필요한, 의도가 불분명한 코드는 최대한 지양해야겠네요!
의도가 분명한 코드를 작성하도록 노력해야겠어요.
감사합니다 :)
```java
public static Long createId() {
    id = id + 1;
    return id;
}
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/controller/TaskController.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<TaskResponseDto>> gets() {
```
할 일 목록을 가져오는 일을 하고 있다보니 `gets`가 된 것 같아요. 더 명확한 이름으로 사용하는게 좋아보입니다.

alex-gibeom replied:
그렇겠네요!
컨트롤러는 습관적으로 그냥 gets, get... 이런식으로 작성했었는데 명확하게 작성하도록 해야겠습니다 :)

![image](https://user-images.githubusercontent.com/59248326/196868545-95e08688-9486-4bdf-93e2-2c42cf64f6b2.png)

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/controller/TaskController.java:
```diff
@@ -0,0 +1,85 @@
+    public ResponseEntity<List<TaskResponseDto>> gets() {
+        List<Task> tasks = taskService.getTasks();
+
+        return ResponseEntity.ok().body(
+                tasks.stream()
+                        .map(TaskResponseDto::from)
+                        .collect(Collectors.toList())
+        );
```
단순히 응답 상태코드를 `200 OK`로 응답하기 위해 `ResponsEntity`를 사용한 것이라면 `@ResponsStatus`를 사용할수도 있겠어요

alex-gibeom replied:
`200 ok` 이외에도 `body()`에 응답 데이터 값을 담는 작업도 진행하도록 하였습니다.
(body() 안에서 Entity -> ResponseDto 변환 작업 같이 수행)

모든 컨트롤러 반환 형식을 일관성 있게 `ResponseEntity.status().body()` 혹은 반환 값이 없다면 `ResponseEntity.status().build()` 로 변경해보았는데 좀 더 좋은 방법이 있을지 궁금합니다!

hannut91 replied:
Spring Web MVC에서는 객체를 반환하면 응답 바디에 들어갑니다. 따라서  `ResponseEntity`를 따로 사용하지 않아도 됩니다.

```java
return tasks.stream()
    .map(TaskResponseDto::from)
    .collect(Collectors.toList());
```

alex-gibeom replied:
아하 그렇군요 감사합니다 :)

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/controller/TaskController.java:
```diff
@@ -0,0 +1,85 @@
+    public ResponseEntity<String> delete(@PathVariable @Min(0) Long id) {
+        taskService.deleteTask(id);
+
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
```
왜 데이터가 없을 때 204로 응답을 해야 하는지 다음 문서를 읽어봅시다. https://tools.ietf.org/html/rfc2616#section-9.7

alex-gibeom replied:
204는 요청에 대한 처리가 완료되었지만, 응답 시 response body에 데이터가 포함되지 않을 경우 사용한다고 합니다.
또한 [mozilla mdn 사이트](https://developer.mozilla.org/ko/docs/Web/HTTP/Status/204)에서는 요청에 성공했으나 클라이언트가 현재 페이지에서 벗어나지 않아도 된다는 것을 나타낸다고도 합니다.

개인적인 견해로는 DELETE는 리소스를 삭제하는 요청이므로, 응답 본문에 반환할 데이터가 없을뿐더러 삭제하였다 하더라도 기존 페이지를 그대로 유지하기 위해 사용되는 것으로 보입니다!
틀린 내용이 있다면 바로 잡아주시면 감사히 학습하겠습니다 :)

ssisksl77 commented:
> 추가로 오늘 작업건에 대해 아래와 같이 설계를 해보았습니다! 피드백 해주시면 감사하겠습니다 😄
> 
> ![image](https://user-images.githubusercontent.com/59248326/196424732-ffdc0e17-5434-495f-b316-b1c9bd22b4ca.png)

와~ 이렇게까지 만드시다니...!!!

ssisksl77 commented:
설계에 대한 것은 코드를 통해서 질문드리도록 할게요~

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/domain/response/TaskResponseDto.java:
```diff
@@ -0,0 +1,19 @@
+
+@Getter
+@Builder
+public class TaskResponseDto {
```
이 코드를 쓰는 곳은 정확이 어느 레이어 일까요?

alex-gibeom replied:
Presentation Layer (Controller)에서 사용되는 객체입니다!
Service -> Controller로 반환된 Entity를 클라이언트로 응답해주기 전에, `TaskResponseDto`로 변환되어 필요한 응답 값만을 반환하는 역할을 수행하도록 만들어보았습니다!

![image](https://user-images.githubusercontent.com/59248326/196854195-5ab71452-97c9-483c-9944-21d6f5bb3d16.png)

ssisksl77 replied:
Controller에서 사용되는 객체라면 controller 패키지에 있는 것이 좋아보입니다.

하지만 이렇게 위치시킨 이유가 있으실거라 생각해요!

그 고민했던 생각을 공유해주세요!

alex-gibeom replied:
저는 task라는 도메인을 담당하는 패키지를 생성하고 그 안에 아래와 같이 배치했었습니다!
> controller, service, repository, domain

![image](https://user-images.githubusercontent.com/59248326/197112427-f69f142c-7b87-42fa-97c7-07c79b54866f.png)

따라서 domain 안에 Entity와 request, response 패키지를 따로 분리해서 담아놨는데요.
해당 패키지 구조는 습관적으로 배치해놓은 것 같습니다 😂

생각해보니 TaskRequestDto와 TaskResponseDto는 controller에서만 사용되니,
멘토님 말씀대로 controller 패키지에 넣는 것이 더 합리적으로 보이네요 :)

---

아래 사진처럼 패키지 구조를 수정하였습니다!

![image](https://user-images.githubusercontent.com/59248326/197121560-180f8eb4-69ea-45ce-9c2a-7b71cf47afbe.png)


한가지 궁금한점은 현재 domain 패키지에 있는 **Entity 클래스인 Task**는 그대로 배치해도 될지,
다른 곳에다 배치하는게 좋을지 고민됩니다!

ssisksl77 replied:
일단 domain에 무엇을 배치한다기보다 service와 domain의 어떤 차이가 있다고 생각하시고 나누셨는지가 궁금해요!

alex-gibeom replied:
제가 생각했던 service와 domain의 차이는 아래와 같습니다!
- service : 특정 주제(Task)에 대한 **비지니스 로직**을 중심으로 담당하는 객체
- domain : 특정 주제(Task)에 대한 **데이터 그 자체**를 담당하는 객체 (Entity 객체와 비슷한 맥락이라고 생각했습니다)

추가적인 건 아래와 같습니다!
- dto : 단순히 프로세스간 데이터를 전달하기 위한 **택배 상자** 같은 객체
- controller : View <-> model을 연결해주는 역할로써 요청 데이터를 통해 응답에 필요한 데이터 처리를 지정해주는 역할을 하는 객체

추가적으로 찾아볼 때 domain은 Entity와 Value로 나누어진다고 하네요..!

생각해보면 정확하게 알고 분리한 것이 아니라, 뭉뚱그려서 알고 있는 상태로 습관적으로 분리한 것 같습니다 😂

ssisksl77 replied:
숨김처리된 코멘트로 답글을 인지하기가 쉽지 않네요 ㅎㅎ

그렇다면 domain은 service 레이어에 속하는 걸까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public Task createTask(TaskRequestDto taskRequestDto) {
+        Task task = taskRequestDto.toEntity(IdGenerator.createId());
+        boolean result = taskRepository.create(task);
+
+        if (!result) {
+            throw new RestApiException(ErrorCode.CONFLICT);
+        }
```
Task를 생성했을 때, 생성이 실패한다면 `result`가 `false`가 된다고 이해했습니다. `result`라는 말은 어떤 말로도 해석이 될 수 있으므로, 더 명확한 이름을 사용하는게 좋아보입니다. `created`같은 이름을 사용해서 이름만 보아도 `boolean`타입의 변수인지 알 수 있도록 하는 것은 어떨가요?

alex-gibeom replied:
변수명을 미처 신경쓰지 못했던 것 같습니다..!
앞으로 사소한 변수명 하나도 신경써서 작성해야겠습니다 :)
감사합니다

```java
boolean isSuccessCreated = taskRepository.create(task);

if (!isSuccessCreated) {
    throw new RestApiException(ErrorCode.PROCESSING_FAIL);
}
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        boolean result = taskRepository.create(task);
+
+        if (!result) {
+            throw new RestApiException(ErrorCode.CONFLICT);
```
이 예외를 왜 던지는지 이해하기 어렵습니다.

alex-gibeom replied:
1주차 멘토링에서 DELETE에 대한 성공/실패 여부를 핸들링 하는 것이 좋다고 하셨는데, 그 때 추가적인 정책이 있다고 가정하고 핸들링 해보라고 하셔서 `409 에러`로 했었습니다.
근데 생각해보니 CREATE에 대한 핸들링도 따로 필요할 것 같아서 기존 `409 에러`를 그대로 사용했는데, 생각해보니 이유가 타당하지 않은 것 같네요..!

`500 에러`로 변경하였는데, 요구 사항이 명확하지 않은 부분에 대해서는 어떤 응답 코드로 반환해야 좋을지 궁금합니다!
```java
if (!isSuccessCreated) {
    throw new RestApiException(ErrorCode.PROCESSING_FAIL);
}

// ErrorCode enum 객체
public enum ErrorCode {
    PROCESSING_FAIL(HttpStatus.INTERNAL_SERVER_ERROR, "서버 오류로 인해 요청이 실패하였습니다");
}
```

hannut91 replied:
일단 `RestApiException`예외 이름이 무엇인지 이해하기 어렵습니다. 예외를 응답할 때는 API 사용자가 어떻게 대응해야하는지가 드러나는게 좋습니다. 

클라이언트가 `서버 오류로 인해 요청이 실패하였습니다` 메시지를 받았다고 가정해 봅시다. 그럼 클라이언트는 어떤 행동을 할 수 있을까요?

alex-gibeom replied:
"잠시 후 다시 요청해주세요. " 같은 대응 가능한 문구를 사용해야겠군요!
감사합니다 :)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/IdGenerator.java:
```diff
@@ -0,0 +1,12 @@
+package com.codesoom.assignment;
+
+public class IdGenerator {
```
유틸 함수를 사용하는 것처럼 사용하는 건가요?

그렇다면 최소한의 동시성을 고려해보면 좋겠어요!

alex-gibeom replied:
Task id를 따로 관리하고, 생성 시 자동으로 증가해주는 역할을 위해 생성하였습니다!
동시성 문제를 깜빡했었네요..!
createdId() 메서드에 synchronized를 추가하였습니다!
```java
public class IdGenerator {
    private static Long id = 0L;
    private IdGenerator() {}

    public static synchronized Long createId() {
        id = id + 1;
        return id;
    }
}
```

추가로 synchronized 키워드를 붙이면 어떤 원리로 동시성 문제가 해결되는지 궁금해서 간단하게 찾아보았습니다!
- synchronized 메서드는 **클래스의 인스턴스 단위로** 동기화가 일어난다 (lock)
- static synchronized 메서드는 해당 **클래스의 객체 단위로** 동기화가 일어난다 (lock)
- id도 같이 static으로 설정함으로써 static 메서드와 동기화가 이루어진다

<br>

[정리]
> static synchronized 동작 방식은 클래스 단위로 lock이 발생하며, 하나의 쓰레드에서 메서드에 접근 중이면 다른 쓰레드가 접근하지 못하도록 lock을 걸고 대키 큐로 보낸 후, 메서드가 종료 될 시 lock을 해제하고 대기 큐의 작업 1개를 실행하는 방식입니다

틀린 부분이 있다면 피드백 부탁드립니다 😄

ssisksl77 replied:
synchronized가 아주 편한 방법이죠! 
이런 기술에 대한 공부 후 설명을 할 때는 꼭 레퍼런스가 있는 것이 좋아요! 
레퍼런스가 블로그라면? 신빙성이 떨어지게 됩니다! 

그러니 synchronized 라는 키워드를 오라클 자바 튜토리얼 같은 문서를 찾아본다면!
오히려 저보다 전문가인 오라클을 통해서 배우게 되시는 것입니다!

IT기술은 외워도 미래엔 많은 것이 바뀔 것 같아요! 그럴수록 남는 것은 공식문서라고 생각해요! 

추가로 synchronized 외에 AtomicLong 같은 객체도 있습니다. 이것도 공부해보면 좋은 경험이실거에요!

alex-gibeom replied:
오 Atomic 클래스는 처음 들어봤습니다!! 감사합니다 :)

멘토님이 공식 문서의 중요성을 일깨워주셔서, 오늘 하루 여러 개념들을 공식 문서로 읽어보았습니다.

AtomicLong도 아래 공식 문서에서 쭉 한번 읽어봤는데요...
- https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/package-summary.html
- https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicLong.html

솔직하게 영어가 익숙하지 않아, 1시간동안 매달려봤는데도 내용이 잘 들어오지 않는 것 같습니다 😂
결국 또 블로그와, 직접 사용해보면서 코드 내부를 디버깅 해보는 식으로 익혔네요..!

```java
private volatile long value;
```

AtomicLong 내부의 value는 **vilatile**로 선언되어있고, 이는 코어가 변수의 값을 읽어올 때 캐시가 아닌 메인 메모리에서 읽어옴으로써 멀티 쓰레드 환경에서 값의 불일치가 해결된다고 Java의 정석 책에 나와있네요!!

<br>

#### [질문 사항]
혹시 공식 문서를 통해 학습하는 노하우 같은 것이 있을까요?
공식 문서 읽는 법을 찾아보니, 이해 안되도 그냥 무조건 읽는게 답이라고 하는데, 혹시 개인적인 노하우가 있으신지 궁금합니다!

ssisksl77 replied:
개인적인 노하우는 저도 없습니다. 
하지만 중요한 것은 공식문서를 추구하다보면 미래에 분명 블로그로 쉽게 얻을 때보다 나은 개발자가 됨을 확신합니다.
저의 경우를 다시 생각해보면 이해가 안되었던 것은 적은 경험때문이지 않았나 싶습니다.

alex-gibeom replied:
잘 새겨 듣고 어렵더라도 공식 문서를 추구해보겠습니다!
감사합니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public Task getTaskById(Long id) {
+        return taskRepository.findById(id)
+                .orElseThrow(() -> new RestApiException(ErrorCode.NOT_FOUND));
```
서비스 레이어가 RestAPI를 통해 호출한 것인지 알아야 할까요?!

alex-gibeom replied:
안그래도 이 부분에 대해서 궁금했었습니다.

현재 제가 설계한 Controller - Service - Repository의 구조는 다음과 같습니다
#### [Controller 역할]
- PathVariable 데이터 검증
- RequestBody 데이터 검증
- **요청 데이터 Service로 전달**
- Service로부터 반환된 Entity -> ResponseDto 변환 및 응답 반환

#### [Service 역할]
- **RequestDto -> Entity 변환**
- 응답 데이터에 필요한 데이터 로직 수행
- 데이터 로직 도중 반환 값 검증 수행
- Controller로 Entity 반환

#### [Repository 역할]
- 넘어온 Entity를 통해 단순 데이터 CRUD 작업 후 데이터 반환

---


기존에는 Controller에서 Entity로 변환 후 Service로 넘겨주는 방식으로 구상했습니다.
이렇게 해야 해당 Service를 여러 곳에서 사용할 수 있으니 말입니다.

하지만 찾아봤던 글 중 해당 내용이 납득이 돼서 Controller의 request 데이터를 그대로  Service에 넘겨주고, Service에서 필요한 Entity를 생성하는것이 좋다고 생각하게 되었습니다.

<img src="https://user-images.githubusercontent.com/59248326/196885200-9011d9d7-281a-4e09-b1e9-7dfd403544e6.png" width="100%">


또한 **개인적인 견해**는 Controller에서 Entity를 만들어주고 Service에 전달하는 경우,
복잡한 도메인에서는 Request 데이터만으로 원하는 Entity를 생성할 수 없고,
여러번의 Service 데이터 처리를 거쳐 원하는 Entity를 생성할 수 있는 경우가 있다고 생각합니다.

하지만 이럴 경우 **Controller에서 Service에 대한 로직들이 발생**하기 때문에 Service 상단 부분에서 RequestDto -> Entity 변환 작업을 하는 것이 **Controller를 깔끔하게 유지**할 수 있다고 생각했습니다!

#### 해당 부분에 대해서 멘토님의 의견을 여쭤보고 싶습니다!
RequestDto -> Entity 변환은 과연 어디서 하는 것이 좋은 방법일까요..!? 😂
참고로 Entity -> 응답 Dto는 Controller에서 진행하도록 하였습니다 :)

ssisksl77 replied:
안녕하세요! 저의 질문에서 많이 확장되었군요!

ssisksl77 replied:
저의 질문은 서비스 레이어에서 RestAPI를 쓰는지 알아야 하는지 입니다.

ssisksl77 replied:
하지만 RequestDto에 대한 질문을 곰곰히 생각해보면, 지금 RequestDto를 너무 범용적으로 사용하는 것이 아닌가 싶습니다!

alex-gibeom replied:
아하 제가 질문을 다르게 인지했나 보네요!
말씀하신 부분이 RestApiException 때문인걸로 인지했는데, 다시 보니 Service에 해당 이름은 어울리지 않다고 느껴지네요..!


해당 부분은 아래와 같이 수정하였습니다 :)
```java
@Override
public Task getTaskById(Long id) {
    return taskRepository.findById(id)
            .orElseThrow(TaskNotFoundException::new);
}
```

alex-gibeom replied:
> 하지만 RequestDto에 대한 질문을 곰곰히 생각해보면, 지금 RequestDto를 너무 범용적으로 사용하는 것이 아닌가 싶습니다!

멘토님이 생각하시기엔, RequestDto와 ResponseDto 모두 Controller에서만 사용하는 것이 더 낫다는 말씀이신가요??

ssisksl77 replied:
기범님이 생각하시기에!
- 꼭 하나만 필요하다면 필수적으로 필요한 컨트롤러에
- 둘다 Dto가 필요한 상황이면 레이어별로 Dto를 만드는 것이 좋겠다는 의견입니다!

alex-gibeom committed:
fix: build.gradle 주석 설명 수정

alex-gibeom committed:
refactor: IdGenerator 메서드 동시성 처리 및 전위 연산자 삭제

alex-gibeom committed:
refactor: Controller 메서드 명 수정 및 Put,Patch 다중 매핑으로 변경

alex-gibeom committed:
refactor: Controller 응답 포맷 일관성 있게 수정

alex-gibeom committed:
refactor: Service 내 create, delete 성공/실패 여부 핸들링 변수명 수정 및 실패 시 500에러로 수정

- (result -> isSuccessCreated)
- (result -> isSuccessDeleted)

alex-gibeom committed:
merge: [Day4] - 백엔드 리팩토링

alex-gibeom committed:
feat: CORS 설정 추가

alex-gibeom commented:
pr 요청 여러번 드려서 죄송합니다 😂
github가 렉걸린줄 알고 여러번 눌렀네욥..!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/controller/TaskController.java:
```diff
@@ -0,0 +1,81 @@
+    public ResponseEntity<List<TaskResponseDto>> getTasks() {
+        List<Task> tasks = taskService.getTasks();
+
+        return ResponseEntity.status(HttpStatus.OK)
+                .body(tasks.stream()
+                        .map(TaskResponseDto::from)
+                        .collect(Collectors.toList())
+                );
```
```suggestion
        return ResponseEntity.status(HttpStatus.OK)
                .body(TaskResponseDto.from(tasks));
```

어떻게 하는지보다 무엇을 하는지가 보이게 코드를 만든다면 가독성이 좋아져요!

alex-gibeom replied:
그러네요!! 해당 부분을 생각하지 못했습니다..!

(6a74532) from 메서드를 오버로딩하여 수정해보았습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/domain/Task.java:
```diff
@@ -0,0 +1,30 @@
+    private final Long id;
+    private final String title;
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Task task = (Task) o;
+        return Objects.equals(id, task.id) && Objects.equals(title, task.title);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, title);
+    }
```
롬복을 좋아하신다면 이것도 롬복으로 대체할 수 있지 않을까요?

alex-gibeom replied:
(111516f) 찾아보니 `@EqualsAndHashCode` 가 있네요 :)

추가로 알아본 결과 `@Data`도 있지만, 해당 어노테이션은 아래와 같은 단점으로 사용을 지양해야된다고 합니다!
- 최대한 지양해야 될 `@Setter`를 생성해줍니다.
- `@AllArgsConstructor`, `@RequiredArgsConstructor` 는 생성자를 생성해주는 도구인데, 
필드 변수의 순서에 영향을 받는 특징이 있어, 순서에 의한 잠재적 오류를 가진다는 단점 때문에 지양해야되는 기능이라고 합니다.

틀린 부분이 있다면 피드백 부탁드립니다 😄

ssisksl77 replied:
맞는말입니다! 
여기서도 지적해드리고 싶은 것은 오히려 '사용을 지양해야된다고 합니다!' 입니다. 카더라의 출처가 있는 것이 좋습니다. 그것이 공식문서에 가까운 것이 중요합니다!

ssisksl77 replied:
즉, 이 경우 어노테이션에 관한 공식문서가 있어서 필드 변수의 순서에 영향을 받는 특징이 있다는 것과 이에 대한 위험성을 토론한 링크라도 있는 것이 좋습니다

alex-gibeom replied:
최대한 공식 문서와 가까운, Reference가 공식 문서인 것을 찾아보니 결국은 github, 영어 블로그네요.. 😂

해당 Lombok 오픈 소스 github에서 관련 주제를 토론한 것을 찾아보았습니다!
https://github.com/projectlombok/lombok/issues/2094

ssisksl77 replied:
이름부터!!! 아주 좋습니다

ssisksl77 replied:
나아가보면 롬복 자체에 대해 많은 의구심을 품는 사람들도 있습니다. 

저는 겪어본 적 없지만 오히려 롬복 때문에 고통을 겪는 경우도 왕왕있는 것 같습니다.

이 점도 한번 고민해보는 것도 좋을 것 같습니다.

ssisksl77 replied:
세상에 답은 항상 변하더라고요. 

롬복이 진리다 엄청난 프로젝트다 라고 말하던 시기가 있던반면 
롬복이 별로다 라고 말하는 사람들도 많아졌습니다.

그리고 이젠 코틀린으로 넘어가면서 롬복이 필요없다고 말씀하시는 분들도 있고요.

스스로 답을 찾아보시면 좋겠습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/service/TaskService.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+
+public interface TaskService {
+    List<Task> getTasks();
+
+    Task getTaskById(Long id);
+
+    Task createTask(TaskRequestDto taskRequestDto);
+
+    Task updateTask(Long id, TaskRequestDto taskRequestDto);
+
+    void deleteTask(Long id);
```
인터페이스에는 항상 주석을 넣어주세요!

alex-gibeom replied:
인터페이스에 주석을 달아야 하는지 몰랐네요!

(f803f07) 해당 수정에서 Javadoc 형식으로 주석을 달아보았습니다
주석을 처음 작성해보는거라 잘 작성했을지 모르겠네요 😓

ssisksl77 replied:
인터페이스에 주석은 필수입니다! 

현재 사용하지는 자바 컬렉션 API나 다른 객체들을 한번 둘러보세요! 인터페이스와 구현체들에 주석들이 꽤나 많습니다!

메소드 외에 인터페이스 자체나 클래스에도 작은 내용이라도 적혀있는 경우가 많아요!

alex-gibeom replied:
그러고보니 저도 자바 API 내부를 까볼때 주석을 해석하곤 했는데, 인지하지 못하고 있었네요!
주석을 마냥 지양해야된다고만 생각하고 있었는데 잘못된 생각이였던 것 같습니다..!

주석 없이도 메서드명을 통해 동작을 유추할 수 있게 명명규칙을 잘 지키는 동시에 적절한 주석도 같이 달아줘야겠군요 :)

alex-gibeom committed:
refactor: Task(Entity 객체)의 equals(), hashCode() Lombok 기능으로 대체

- `@EqualsAndHashCode`로 적용

alex-gibeom committed:
refactor: dto 파일 구조 변경 - controller 패키지 내부로 이동

alex-gibeom committed:
refactor: 예외 계층 구조 수정 및 에러 클래스 변경

- commonException (공통 부모)
   - 직접 생성할 Exception들 (자식)

alex-gibeom committed:
refactor: TaskResponseDto 객체의 반환 타입 List인 from 메서드 추가 (오버로딩)

alex-gibeom committed:
refactor: TaskService 인터페이스 주석 추가

alex-gibeom committed:
merge: [Day5] - 백엔드 리팩토링

alex-gibeom commented:
(2ceed9f) 공통 에러 핸들링 구조를 수정한 후 구조도를 다시 그려봤습니다 :)
먼가 저만의 틀을 만들어가는게 재밌네요ㅎㅎ


![image](https://user-images.githubusercontent.com/59248326/197197023-22abea54-eae5-42a3-a54f-e2566854d32f.png)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/domain/IdGenerator.java:
```diff
@@ -0,0 +1,13 @@
+    private IdGenerator() {
+    }
+
+    public static synchronized Long createId() {
```
주석이 있으면 좋을 것 같아요!

alex-gibeom replied:
(dcbe40f) 해당 메서드의 주석도 추가해보았습니다!

alex-gibeom replied:
한가지 궁금한 점은 보통 Util성 클래스, 메서드에는 항상 주석을 달아주는 것이 좋은 건가요?
주석을 달아야하는 메서드와 클래스의 기준이 궁금합니다!

평소에 일반 Controller, Service 구현체 등에는 오히려 주석이 가독성을 해치기 때문에 명명규칙을 잘 지켜야하는 것으로 알고 있어서요!

ssisksl77 replied:
항상. 항상 좋습니다. 
안 달아서 좋다? 그런 경우는 없습니다.

명명규칙은 네이밍규칙을 말하시는 거죠. 

주석이 없어도 될만한 이름을 짓는 것과 주석을 붙이는 것과는 다르다고 생각합니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,51 @@
+@Builder
+@Slf4j
+public class ErrorResponse {
+    private final LocalDateTime timestamp = LocalDateTime.now();
```
이 timestamp는 정확히 무엇에 대한 시간일까요?

alex-gibeom replied:
에러가 발생한 시점의 시간을 표현하려고 하였습니다.
지금 보니 변수명이 애매모호한 것 같습니다..!

(10844b1) 아래와 같이 수정하였습니다!
```java
public class ErrorResponse {
    private final LocalDateTime ERROR_OCCURED_TIME = LocalDateTime.now();
    private final String ERROR_MESSAGE;
}
```

의도와 다른 대답이었다면 말씀해주시면 감사하겠습니다!!

ssisksl77 replied:
예외가 생성된 시간이 필요하다면 예외가 생성되는 객체에 있는 것이 좋을 것 같아요!

ssisksl77 replied:
혹은 예외마저 날짜를 인자로 받는 것을 고려해볼 수 있겠어요!

alex-gibeom replied:
그게 더 자연스러워 보이겠네요
감사합니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/task/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public Task createTask(TaskRequestDto taskRequestDto) {
+        Task task = taskRequestDto.toEntity(IdGenerator.createId());
```
taskRequestDto는 어느 레이어에 위치한 객체인가요?

alex-gibeom replied:
`taskRequestDto` 또한 Presentation Layer(Controller)에서 Service Layer로 데이터 값을 전달해주기 위한 객체로 사용하였습니다!

따라서 `taskRequestDto`의 위치는 Presentation Layer(Controller)인 것 같습니다!

alex-gibeom committed:
refactor: IdGenerator AtomicLong 적용

alex-gibeom committed:
refactor: Javadoc주석 수정 (TaskService, TaskIdGenerator)

alex-gibeom committed:
refactor: ErrorResponse 에러 발생 시간을 위한 변수명 수정

alex-gibeom committed:
merge: [Day6] - 백엔드 리팩토링

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/exception/ErrorResponse.java:
```diff
@@ -15,24 +15,24 @@
     public static ErrorResponse from(final CommonException exception) {
         return ErrorResponse.builder()
-                .message(exception.getMessage())
+                .ERROR_MESSAGE(exception.getMessage())
```
https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

자바 네이밍 컨벤션을 참고해주세요!

alex-gibeom replied:
static final (클래스 상수) 일 때만 대문자에 _를 붙이는 거였군요..!
해당 변수는 인스턴스 상수이므로 카멜케이스로 수정해야겠군요 :)
다시 수정하였습니다!

```java
public class ErrorResponse {
  private final LocalDateTime occuredTime;
  private final String message;
}
```

ssisksl77 replied:
좋습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static ErrorResponse from(HttpServletRequest request, final Exception exception) {
+        log.error("[ERROR-]\t{}\t{}\t{}", request.getMethod(), request.getRequestURI(), 500, exception.getStackTrace());
```
리턴을 하는 것도 아니고, 로그를 기록하는 것인데 500이 필요한가 싶습니다!

alex-gibeom replied:
해당 예외 핸들러는 로그를 남기면서, exception의 getMessage만 반환하도록 설정하였습니다..!
```java
public static ErrorResponse from(HttpServletRequest request, final Exception exception) {
    log.error("[ERROR-]\t{}\t{}\t{}", request.getMethod(), request.getRequestURI(), 500, exception.getStackTrace());
    log.error("{}", exception.getStackTrace());

    return ErrorResponse.builder()
            .message(exception.getMessage())
            .build();
}
```


그냥 500을 반환해보면 아래와 같이 반환 값들이 엄청나게 복잡하길래, 보통 어떻게 처리하는지 몰라서 이렇게 처리해보았습니다..!
<img src="https://user-images.githubusercontent.com/59248326/197390828-022abec5-4733-4840-971c-5596f487259e.png" width="30%">

<br>

보통은 어떤식으로 처리하는지 간단하게 알려주실 수 있으신가요?

ssisksl77 replied:
IDE를 쓰신다면 breakpoint를 써보는 방법이 있을 것입니다!

ssisksl77 replied:
질문이 모호했나보군요. 로그를 기록할 때 500이라는 상수가 필요한가입니다

ssisksl77 replied:
추가로 지금 리턴하는 방식에 문제는 없습니다!

alex-gibeom replied:
아하 제가 잘 못 이해한 것 같습니다!
500을 남긴 이유는 아래와 같은 형태로 로그를 남겨보고 싶었습니다.
![image](https://user-images.githubusercontent.com/59248326/197394240-6e911889-40d2-4cbf-9c40-50a4bf9d07dc.png)

하지만 생각해보니 굳이 `Exception.class` 자체를 핸들링하는데 500을 남길 필요가 없을 것 같기도 하고,
또한 만약 필요하다 하더라도 저런식으로 하드코딩은 좋지 못할 것 같다라는 생각이 드는 것 같습니다..!

ssisksl77 commented:
안녕하세요! 
꼭 이번주가 끝이라고 리뷰가 끝이라고 생각하시지 마시고, 질문을 계속해서 이어갈 수 있습니다. 
혹은 질문을 이어서 week3으로 링크를 가져와서 이어가도 되고요!

alex-gibeom committed:
refactor: ErrorResponse 변수명 수정

alex-gibeom committed:
refactor: 실제 예외 생성 객체에서 발생 시간 생성 하도록 변경(CommonException)

alex-gibeom commented:
> 안녕하세요! 꼭 이번주가 끝이라고 리뷰가 끝이라고 생각하시지 마시고, 질문을 계속해서 이어갈 수 있습니다. 혹은 질문을 이어서 week3으로 링크를 가져와서 이어가도 되고요!

아 해당 주가 끝나도 리뷰가 가능하군요..!
OT에서 들은 바로는 그 주의 리뷰는 그 주 넘어서는 피드백이 불가능하다고 전달받아서, 꼭 일요일까지 해당 주차 리뷰를 전부 받아야되는 줄 알았습니다!
해당 주가 지나면 기회가 없는 건줄 알았네요 😂

감사합니다!

ssisksl77 commented:
> > 안녕하세요! 꼭 이번주가 끝이라고 리뷰가 끝이라고 생각하시지 마시고, 질문을 계속해서 이어갈 수 있습니다. 혹은 질문을 이어서 week3으로 링크를 가져와서 이어가도 되고요!
> 
> 아 해당 주가 끝나도 리뷰가 가능하군요..! OT에서 들은 바로는 그 주의 리뷰는 그 주 넘어서는 피드백이 불가능하다고 전달받아서, 꼭 일요일까지 해당 주차 리뷰를 전부 받아야되는 줄 알았습니다! 해당 주가 지나면 기회가 없는 건줄 알았네요 😂
> 
> 감사합니다!

코드를 바꾸는 것을 계속 추적하는 것은 어렵지만 질문에 대한 답변은 이어서 할 수 있는 부분이지 않을까 싶어요! 
일주일 정도가 되면 코드에서 크게 리뷰할 부분이 없어집니다!
그러니 다음 주차에도 완료되지 않은 질문이나, 답변을 했지만 추가 질문이 있으신경우 다음주차 PR에 링크를 함께 올려주시면 편하게 리뷰를 할 수 있을 것 같아요!

alex-gibeom commented:
> 일단 domain에 무엇을 배치한다기보다 service와 domain의 어떤 차이가 있다고 생각하시고 나누셨는지가 궁금해요!

>> 제가 생각했던 service와 domain의 차이는 아래와 같습니다!
>> 
>> * service : 특정 주제(Task)에 대한 **비지니스 로직**을 중심으로 담당하는 객체
>> * domain : 특정 주제(Task)에 대한 **데이터 그 자체**를 담당하는 객체 (Entity 객체와 비슷한 맥락이라고 생각했습니다)
>> 
>> 추가적인 건 아래와 같습니다!
>> 
>> * dto : 단순히 프로세스간 데이터를 전달하기 위한 **택배 상자** 같은 객체
>> * controller : View <-> model을 연결해주는 역할로써 요청 데이터를 통해 응답에 필요한 데이터 처리를 지정해주는 역할을 하는 객체
>> 
>> 추가적으로 찾아볼 때 domain은 Entity와 Value로 나누어진다고 하네요..!
>> 생각해보면 정확하게 알고 분리한 것이 아니라, 뭉뚱그려서 알고 있는 상태로 습관적으로 분리한 것 같습니다 😂

>>> 숨김처리된 코멘트로 답글을 인지하기가 쉽지 않네요 ㅎㅎ 
>>> 그렇다면 domain은 service 레이어에 속하는 걸까요?

<br>

제가 생각했을 때는 domain은 service와 같이 Business Layer에 속하는 것으로 알고 있었습니다..!

mvc를 공부했을 때 아래의 그림으로 공부를 했었습니다!
<img src="https://user-images.githubusercontent.com/59248326/197393886-b8b2d9e9-63d9-4c2e-adad-8ea43b047502.png" width="50%">

근데 선뜻 얘기를 못하는거 보면은 대강만 알고 있었던 것 같네요..
공부가 필요해 보입니다

ssisksl77 commented:
> > 일단 domain에 무엇을 배치한다기보다 service와 domain의 어떤 차이가 있다고 생각하시고 나누셨는지가 궁금해요!
> 
> > > 제가 생각했던 service와 domain의 차이는 아래와 같습니다!
> > > 
> > > * service : 특정 주제(Task)에 대한 **비지니스 로직**을 중심으로 담당하는 객체
> > > * domain : 특정 주제(Task)에 대한 **데이터 그 자체**를 담당하는 객체 (Entity 객체와 비슷한 맥락이라고 생각했습니다)
> > > 
> > > 추가적인 건 아래와 같습니다!
> > > 
> > > * dto : 단순히 프로세스간 데이터를 전달하기 위한 **택배 상자** 같은 객체
> > > * controller : View <-> model을 연결해주는 역할로써 요청 데이터를 통해 응답에 필요한 데이터 처리를 지정해주는 역할을 하는 객체
> > > 
> > > 추가적으로 찾아볼 때 domain은 Entity와 Value로 나누어진다고 하네요..!
> > > 생각해보면 정확하게 알고 분리한 것이 아니라, 뭉뚱그려서 알고 있는 상태로 습관적으로 분리한 것 같습니다 😂
> 
> > > > 숨김처리된 코멘트로 답글을 인지하기가 쉽지 않네요 ㅎㅎ
> > > > 그렇다면 domain은 service 레이어에 속하는 걸까요?
> 
> 
> 제가 생각했을 때는 domain은 service와 같이 Business Layer에 속하는 것으로 알고 있었습니다..!
> 
> mvc를 공부했을 때 아래의 그림으로 공부를 했었습니다! <img alt="" width="50%" src="https://user-images.githubusercontent.com/59248326/197393886-b8b2d9e9-63d9-4c2e-adad-8ea43b047502.png">
> 
> 근데 선뜻 얘기를 못하는거 보면은 대강만 알고 있었던 것 같네요.. 공부가 필요해 보입니다

Domain 객체에 대해서 아래 링크를 보고 고민해보는 시간을 가져보세요!
도메인에 대한 공식적인 문헌은 아니지만 저는 공감이 되는 내용이에요!
https://stackoverflow.com/questions/21339657/whats-the-difference-between-service-layer-and-domain-model-layer

ssisksl77 reviewed (approved):
한 주간 고생하셨습니다!
